nimare/annotate/gclda.py

```diff
@@ -14,6 +15,247 @@
+
+    # Iterate over all peaks x, and sample a new y and r assignment for each
+    for i_ptoken, doc in enumerate(doc_idx):
+        topic = peak_topic_idx[i_ptoken]
+        region = peak_region_idx[i_ptoken]
+
+        # Decrement count-matrices to remove current ptoken_topic_idx
+        # because ptoken will be reassigned to a new topic
+        ny_region_topic[region, topic] -= 1  # Decrement count in Subregion x Topic count matx
+        ny_doc_topic[doc, topic] -= 1  # Decrement count in Document x Topic count matrix
+
+        # Retrieve the probability of generating current x from all
+        # subregions: [R x T] array of probs
+        p_x_subregions = (peak_probs[i_ptoken, :, :]).transpose()
+
+        # Compute the probabilities of all subregions given doc
+        #     p(r|d) ~ p(r|t) * p(t|d)
+        # Counts of subregions per topic + prior: p(r|t)
+        p_region_g_topic = ny_region_topic + delta
+
+        # Normalize the columns such that each topic's distribution over subregions sums to 1
+        p_region_g_topic_sum = np.sum(p_region_g_topic, axis=0)
+        p_region_g_topic = np.divide(p_region_g_topic, p_region_g_topic_sum)
+
+        # Counts of topics per document + prior: p(t|d)
+        p_topic_g_doc = ny_doc_topic[doc, :] + alpha
+
+        # Reshape from (ntopics,) to (nregions, ntopics) with duplicated rows
+        p_topic_g_doc = p_topic_g_doc.repeat(n_regions).reshape((-1, n_regions)).T
+
+        # Compute p(subregion | document): p(r|d) ~ p(r|t) * p(t|d)
+        # [R x T] array of probs
+        p_region_g_doc = p_topic_g_doc * p_region_g_topic
+
+        # Compute the multinomial probability: p(z|y)
+        # Need the current vector of all z and y assignments for current doc
+        # The multinomial from which z is sampled is proportional to number
+        # of y assigned to each topic, plus constant gamma
+        # Compute the proportional probabilities in log-space
+        logp = nz_doc_topic[doc, :] * np.log(
+            (ny_doc_topic[doc, :] + gamma + 1) / (ny_doc_topic[doc, :] + gamma)
+        )
+        # Add a constant before exponentiating to avoid any underflow issues
+        p_peak_g_topic = np.exp(logp - np.max(logp))
+
+        # Reshape from (ntopics,) to (nregions, ntopics) with duplicated rows
+        p_peak_g_topic = p_peak_g_topic.repeat(n_regions).reshape((-1, n_regions)).T
+
+        # Get the full sampling distribution:
+        # [R x T] array containing the proportional probability of all y/r combinations
+        probs_pdf = p_x_subregions * p_region_g_doc * p_peak_g_topic
+
+        # Convert from a [R x T] matrix into a [R*T x 1] array we can sample from
+        probs_pdf = np.reshape(probs_pdf, (n_regions * n_topics))
+
+        # Normalize the sampling distribution
+        probs_pdf = probs_pdf / np.sum(probs_pdf)
+        # Float precision issue in Numba: https://github.com/numba/numba/issues/3426
+        probs_pdf = np.trunc(probs_pdf * (10**12)) / (10**12)
```

Comment:
This line will need to be removed in the future. It is ensuring the sum never goes over 1.0 so that we don't get this error in numba, but this truncation is causing the maps on the [example](https://nimare--744.org.readthedocs.build/en/744/auto_examples/03_annotation/04_plot_gclda.html#look-at-topics) to look different to the ones from previous versions.

Reply:
I was running some tests with Neurosynth and Neuroquery, and got the error again:

```
File "<...>/lib/python3.9/site-packages/nimare/annotate/gclda.py", line 622, in fit
    self._update(loglikely_freq=loglikely_freq)
  File "<...>/lib/python3.9/site-packages/nimare/annotate/gclda.py", line 681, in _update
    ) = _update_peak_assignments(
  File "<...>/lib/python3.9/site-packages/numba/cpython/randomimpl.py", line 833, in binomial_impl
    raise ValueError("binomial(): p outside of [0, 1]")
ValueError: binomial(): p outside of [0, 1]
```

Reply:
would np.clip help here? 
during normalization, you could do this:
`probs_pdf = (probs_pdf / np.sum(probs_pdf)).clip(0, 1)`

Reply:
I tried using `clip` but it still raised the same error. I think it has to do with the way `numba` has the sum implemented.